solidity/test/FullyBackedBondingTest.js

```diff
@@ -0,0 +1,451 @@
+      )
+
+      await bonding.authorizeOperatorContract(operator, bondCreator, {
+        from: authorizer,
+      })
+
+      await bonding.authorizeSortitionPoolContract(operator, sortitionPool, {
+        from: authorizer,
+      })
+
+      expect(
+        await bonding.availableUnbondedValue(
+          operator,
+          bondCreator,
+          sortitionPool
+        )
+      ).to.eq.BN(expectedFinalBalance, "invalid final available unbonded value")
+    })
+  })
+
+  describe("withdraw", async () => {
+    const value = new BN(1000)
+
+    let initialDeposit
+    let delegationLockPeriod
+
+    beforeEach(async () => {
+      initialDeposit = minimumDelegationValue
+      delegationLockPeriod = await bonding.DELEGATION_LOCK_PERIOD.call()
+
+      await bonding.delegate(operator, beneficiary, authorizer, {
+        from: owner,
+        value: initialDeposit,
+      })
+
+      await bonding.authorizeOperatorContract(operator, bondCreator, {
+        from: authorizer,
+      })
+
+      await time.increase(delegationLockPeriod.addn(1))
+    })
+
+    it("can be called by operator", async () => {
+      await bonding.withdraw(value, operator, {from: operator})
+      // ok, no reverts
+    })
+
+    it("can be called by delegation owner", async () => {
+      await bonding.withdraw(value, operator, {from: owner})
+      // ok, no reverts
+    })
+
+    it("cannot be called before delegation lock period passes", async () => {
+      const operator2 = await web3.eth.personal.newAccount("pass")
+
+      await bonding.delegate(operator2, beneficiary, authorizer, {
+        from: owner,
+        value: initialDeposit,
+      })
+
+      await bonding.authorizeOperatorContract(operator2, bondCreator, {
+        from: authorizer,
+      })
+
+      await time.increase(delegationLockPeriod)
+
+      await expectRevert(
+        bonding.withdraw(value, operator2, {from: owner}),
+        "Delegation lock period has not passed yet"
+      )
+    })
+
+    it("cannot be called by authorizer", async () => {
+      await expectRevert(
+        bonding.withdraw(value, operator, {from: authorizer}),
+        "Only operator or the owner is allowed to withdraw bond"
+      )
+    })
+
+    it("cannot be called by beneficiary", async () => {
+      await expectRevert(
+        bonding.withdraw(value, operator, {from: beneficiary}),
+        "Only operator or the owner is allowed to withdraw bond"
+      )
+    })
+
+    it("cannot be called by third party", async () => {
+      const thirdParty = accounts[7]
+
+      await expectRevert(
+        bonding.withdraw(value, operator, {from: thirdParty}),
+        "Only operator or the owner is allowed to withdraw bond"
+      )
+    })
+
+    it("transfers unbonded value to beneficiary", async () => {
+      const expectedUnbonded = initialDeposit.sub(value)
+
+      const expectedBeneficiaryBalance = web3.utils
+        .toBN(await web3.eth.getBalance(beneficiary))
+        .add(value)
```

Comment:
Can we move those declarations to be next to assertions? It took me a couple of minutes to understand what `.sub(value)` is doing in `expectedUnbonded` declaration.

I'd like to see `await bonding.withdraw(value, operator, {from: operator})` first and then the expected values.

Reply:
We usually first define expectations, next execute test action and then run verification. I can move pieces here as you wish if this helps with understanding the test case.

Reply:
We cannot move `expectedBeneficiaryBalance` after `withdraw` as we have to call `web3.eth.getBalance` to get initial account balance before calling the action changing the state.